contract_isp_invoice/contract.py

```diff
@@ -204,140 +204,443 @@ class account_journal(orm.Model):
+                                           end_month_days),
+                    ))
+
+        return start_date, end_date, ptx
+
+
 class account_analytic_account(orm.Model):
     _inherit = "account.analytic.account"
 
     _columns = {
         'close_date': fields.datetime('Close date'),
-        'close_reason': fields.text('Reasons')
+        'close_reason': fields.text('Reasons'),
     }
 
     _defaults = {
         'close_date': fields.datetime.now
     }
 
-    def create_invoice(self, cr, uid, ids, prorata=False, context=None):
-        return_int = False
-        if isinstance(ids, int):
-            return_int = True
+    def get_last_invoice_date(self, cr, uid, ids, source_process, inv=False,
+                              context=None):
+        res = {}
+        source_process = tuple((source_process or '').split(","))
+        account_ids = tuple(ids)
+        if not ids:
+            return res
+
+        cr.execute(
+            """
+            SELECT account_analytic_line.account_id
+                 , DATE(MAX(account_invoice.date_invoice))
+            FROM account_analytic_line
+            JOIN account_invoice
+              ON account_analytic_line.invoice_id = account_invoice.id
+            WHERE account_analytic_line.account_id IN %s
+              AND account_analytic_line.invoice_id IS NOT NULL
+              AND {neg} COALESCE(account_invoice.source_process, '') IN %s
+            GROUP BY account_analytic_line.account_id
+            """.format(neg=('NOT' if inv else '')),
+            (account_ids, source_process),
+        )
+
+        for account_id, lid in cr.fetchall():
+            res[account_id] = lid
+        return res
+
+    def get_last_invoice_date_non_prorata(self, cr, uid, ids, context=None):
+        return self.get_last_invoice_date(cr, uid, ids,
+                                          PROCESS_PRORATA, inv=True,
+                                          context=context)
+
+    def send_email_contract_invoice(self, cr, uid, ids, context=None):
+        context = context or {}
+
+        if not isinstance(ids, list):
             ids = [ids]
 
-        account_analytic_account_obj = self.pool.get('account.analytic.account')
-        account_analytic_line = self.pool.get('account.analytic.line')
-        contract_service_obj = self.pool.get('contract.service')
-        res_company_obj = self.pool.get('res.company')
         account_invoice_obj = self.pool.get('account.invoice')
+        mail_template_obj = self.pool.get('email.template')
+        ir_model_data_obj = self.pool.get('ir.model.data')
+        mail_template_id = ir_model_data_obj.get_object_reference(
+            cr, uid, 'account',
+            'email_template_edi_invoice')[1]
+        mail_mail_obj = self.pool.get('mail.mail')
+
+        for inv in ids:
+            _logger.info(
+                "Mailing invoice %s" % account_invoice_obj.browse(
+                    cr, uid, inv, context=context).name)
```

Comment:
for logging, I'd just log the id of the invoice rather than have a call to browse for each invoice. 

You should change all you calls to logging methods not *not* use %, but additional arguments. logging is smart enough to not do the costly parameter substitution when the logging is not to be issues (e.g. because the level is too low)